### DISCUSSION TITLE
Compiler testing

### DISCUSSION THREAD

**Discussion by crcrewso** [nrc-cnrc/EGSnrc#680] [Programming]
With LLVM soon to support Fortran, I thought it might be a good idea to create a discussion thread around current and future compilers for future support

Two points of interest to start this off:

- Current Families I have tested
  - GCC
  - PGI
  - Intel
- Upcoming 
  - LLVM and it's derivatives
  - Whatever comes out with Intel's OneAPI

Secondly, I propose that we start dropping official MacOS support, and migrate Windows support to only WSL. 

- WSL performance and 'pause' issues have  been sorted, it can be treated as equivalent to Ubuntu/Fedora/Debian/RedHat/SUSE/other distro of choice. 
  - Installation of LLVM and gcc families is directly supported through each distributions package channel. 
  - GUI passthrough is well documented, well supported, and future usability improvements are road mapped by Microsoft directly (wayland support) 
- Mac has not updated GCC in forever, it's still 4.2 and their branch of LLVM might never support Fortran.

**Comment by ftessier** [comment]
I would be most interested in any report of trying to compile the EGSnrc code base using `Flang` (https://github.com/llvm/llvm-project/tree/main/flang)!

**Reply by crcrewso** [reply]
I've been trying recently. Flang is not yet mainlined, so my testing has not been fruitful. I'm hoping as LLVM 13 gets wider adoption (and LLVM 14) that I can make progress on it. 

oneAPI (replacement iFort) is giving me some issues with the new flags but I'm hopeful on that one too. 

Did you ever find your alternate, simplified configuration scripts? They may speed up my test loops.

**Reply by ftessier** [reply]
I am not sure what you are referring to, or what I promised. 😄 

There is an automated configuration script [`configure.expect`](https://github.com/nrc-cnrc/EGSnrc/blob/master/HEN_HOUSE/scripts/configure.expect) based on the `expect` tool, which will configure EGSnrc automatically. You can edit the script to specify the answers to the configuration options. Typical usage:

```
HEN_HOUSE/scripts/configure.expect a6fc389 3
```

where `a6fc389` is the name you want to give to the configuration (best practice is to set it to the current commit short hash), and `3` is the answer to the `egs_home` compilation question `Choose 1 (all), 2 (some) or 3 (none):`. Select none (3) to check compilation of the EGSnrc core only, or all (1) for a more thorough check!

By the way, you may be interested in the new instructions for [managing EGSnrc installation and upgrade with git.](https://github.com/nrc-cnrc/EGSnrc/wiki/Upgrading-EGSnrc). Following from that, I now use git branches for every separate project, thus ensuring that each one has a clean change log trail leading back to an official release. Perhaps you can use this likewise to test and/or maintain multiple configurations? The only part missing to manage configurations under git is a [post-checkout hook](https://git-scm.com/docs/githooks#_post_checkout) to change the environment variables when checking out branches.
